File: 1268-search-suggestions-system/1268-search-suggestions-system.py
class Trie:
    def __init__(self):
        self.root = TrieNode()


    def insert(self, word):
        node = self.root

        for letter in word:
            target = ord(letter) - ord('a')
            if node.children[target] is None:
                node.children[target] = TrieNode()
            node = node.children[target]
            if len(node.words) < 3:
                node.words.append(word)


    def search(self, word):
        node = self.root

        for letter in word:
            target = ord(letter) - ord('a')
            if node.children[target] is None:
                return list()
            node = node.children[target]

        return node.words


class TrieNode:
    def __init__(self):
        self.children = [None for i in range(26)]
        self.words = list()


class Solution:
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        products.sort()
        trie = Trie()
        result = list()

        for product in products:
            trie.insert(product)

        for i in range(len(searchWord)):
            result.append(trie.search(searchWord[:i+1]))

        return result

File: 1268-search-suggestions-system/1268-search-suggestions-system.java
class Trie {
    TrieNode root;

    public Trie() {
        this.root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode node = this.root;
    
        for (int i = 0; i < word.length(); i++) {
            int target = word.charAt(i) - 'a';
            if (node.children[target] == null) {
                node.children[target] = new TrieNode();
            }
            node = node.children[target];
            if (node.words.size() < 3) {
                node.words.add(word);
            }
        }
    }

    public List<String> search(String word) {
        TrieNode node = this.root;

        for (int i = 0; i < word.length(); i++) {
            int target = word.charAt(i) - 'a';
            if (node.children[target] == null) {
                return new ArrayList<>();
            }
            node = node.children[target];
        }

        return node.words;
    }
}

class TrieNode {
    TrieNode[] children;
    List<String> words;

    public TrieNode() {
        this.children = new TrieNode[26];
        this.words = new ArrayList<>();
    }
}

class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Arrays.sort(products);
        Trie trie = new Trie();
        List<List<String>> result = new ArrayList<>();

        for (String product: products) {
            trie.insert(product);
        }

        for (int i = 0; i < searchWord.length(); i++) {
            result.add(trie.search(searchWord.substring(0, i + 1)));
        }

        return result;
    }
}